monitoring/garden_collector.go

```diff
@@ -0,0 +1,120 @@
+package monitoring
+
+import (
+	"sync"
+
+	"github.com/lightninglabs/taproot-assets/tapgarden"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// assetBalancesCollector is a Prometheus collector that exports the balances
+// of all taproot assets.
+type gardenCollector struct {
+	collectMx sync.Mutex
+
+	cfg      *PrometheusConfig
+	registry *prometheus.Registry
+
+	pendingBatches   *prometheus.GaugeVec
+	completedBatches prometheus.Gauge
+}
+
+func newGardenCollector(cfg *PrometheusConfig,
+	registry *prometheus.Registry) *gardenCollector {
+
+	return &gardenCollector{
+		cfg:      cfg,
+		registry: registry,
+		pendingBatches: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "mint_batches",
+				Help: "Batched mint transactions",
+			},
+			[]string{"batch_pubkey"},
+		),
+		completedBatches: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: "completed_batches",
+				Help: "Total number of completed mint batches",
+			}),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once the
+// last descriptor has been sent.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *gardenCollector) Describe(ch chan<- *prometheus.Desc) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	a.pendingBatches.Describe(ch)
+	a.completedBatches.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *gardenCollector) Collect(ch chan<- prometheus.Metric) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	if a.cfg == nil {
+		log.Error("cfg is nil")
+		return
+	}
+
+	if a.cfg.AssetStore == nil {
+		log.Error("asset store is nil")
+		return
+	}
+
+	a.completedBatches.Set(0)
+
+	// Get the number of pending batches.
+	batches, err := a.cfg.AssetMinter.ListBatches(nil)
+	if err != nil {
+		log.Errorf("unable to list batches: %v", err)
+		return
+	}
+
+	completed := 0
+
+	for _, batch := range batches {
+		state := batch.State()
+
+		switch {
+		case state == tapgarden.BatchStatePending ||
+			state == tapgarden.BatchStateFrozen ||
+			state == tapgarden.BatchStateCommitted ||
+			state == tapgarden.BatchStateBroadcast ||
+			state == tapgarden.BatchStateConfirmed:
```

Comment:
Why not just check for `BatchStatePending`? (and remove the `if` below).

Reply:
So the gauge value may only be updated while state is `BatchStatePending`. The rest of the states are not final, and they may not change the value of the gauge so I need to just register them without changing the gauge (I can't do `len(batch.Seedlings)` again as the array is emptied). The reason I register them is because these states mean that the batch is still "in-flight"